camviewer_ui_impl.py

```diff
@@ -1151,30 +1151,58 @@ def shutdown(self):
-                self.cwd,
+                os.path.expanduser("~"),
                 "Images (*.npy *.jpg *.png *.bmp *.pgm *.tif)",
             )[0]
-            if fileName == "":
+            if filename == "":
                 raise Exception("No File Name Specified")
-            if fileName.lower().endswith(".npy"):
-                np.save(fileName, self.image)
+            if filename.lower().endswith(".npy"):
+                np.save(filename, self.image)
                 QMessageBox.information(
                     self,
                     "File Save Succeeded",
-                    "Image has been saved as a numpy file: %s" % (fileName),
+                    "Image has been saved as a numpy file: %s" % (filename),
                 )
-                print("Saved to a numpy file %s" % (fileName))
-            else:
-                self.ui.display_image.image.save(fileName, format=None, quality=-1)
+                print("Saved to a numpy file %s" % (filename))
+            elif self.ui.display_image.image.save(filename, format=None, quality=-1):
+                # QImage.save returned True, so the save succeeded.
                 QMessageBox.information(
                     self,
                     "File Save Succeeded",
-                    "Image has been saved to an image file: %s" % (fileName),
+                    "Image has been saved to an image file: %s" % (filename),
+                )
+                print("Saved to an image file %s" % (filename))
+            else:
+                # QImage.save returned False, so the save failed.
+                # Check for obvious errors, then retry the save
+                # File already exists? This shouldn't happen due to the qt file dialog.
+                if os.path.exists(filename):
+                    reason = (
+                        f"{filename} already exists! Please pick a different filename"
+                    )
+                # No write permissions?
+                else:
+                    # This is the most robust way to check if we have permissions
+                    # os.access exists but doesn't quite work fully on NFS filesystems as per the Python docs
+                    try:
+                        with open(filename, "w"):
+                            ...
+                    except PermissionError:
+                        reason = f"No permissions to write to {filename}! Please pick a different location."
+                    else:
+                        # Clean up the stub file we made
+                        os.remove(filename)
+                        # I guess we have no idea what went wrong
+                        reason = "Unknown failure! Please try again."
+                QMessageBox.warning(
+                    self,
+                    "File Save Failed",
+                    reason,
                 )
-                print("Saved to an image file %s" % (fileName))
+                return self.onfileSave()
```

Comment:
are we recursing on purpose here?

Reply:
Re-starting the procedure is intended, the recursive call is the easiest way to do this.
Probably re-emitting a signal is the cleaner way.

Reply:
Resolved via swapping out recursive call with a signal